include/net/azure_iot_hub.h

```diff
@@ -0,0 +1,288 @@
+	/** Pointer to data sent/received from the Azure IoT Hub broker. */
+	char *ptr;
+	/** Length of data. */
+	size_t len;
+	/** Quality of Service for the message. */
+	enum mqtt_qos qos;
+	/** Message id used for the message, used to match acknowledgments. */
+	uint16_t message_id;
+	/** Duplicate flag. 1 indicates the message is a retransmission,
+	 *  Usually triggered by missing publication acknowledgment.
+	 */
+	uint8_t dup_flag;
+	/** Retain flag. 1 indicates to the IoT hub that the message should be
+	 *  stored persistently.
+	 */
+	uint8_t retain_flag;
+};
+
+/** @brief Azure IoT Hub direct method data. */
+struct azure_iot_hub_method {
+	/** Method name, null-terminated string. */
+	const char *name;
+	/** Method request ID. */
+	uint32_t rid;
+	/** Method payload. */
+	const char *payload;
+	/** Method payload length. */
+	size_t payload_len;
+};
+
+/** @brief Azure IoT Hub result structure.
+ *
+ *  @details Used to signal result of direct method execution from device to
+ *	     cloud, and to receive result of device twin updates (twin updates
+ *	     sent from the device will receive a result message back from the
+ *	     cloud with success or failure).
+ */
+struct azure_iot_hub_result {
+	/** Request ID to which the result belongs. */
+	uint32_t rid;
+	/** Status code. */
+	uint32_t status;
+	/** Result payload. */
+	const char *payload;
+	/** Size of payload. */
+	size_t payload_len;
+};
+
+/** @brief Struct with data received from Azure IoT Hub broker. */
+struct azure_iot_hub_evt {
+	/** Type of event. */
+	enum azure_iot_hub_evt_type type;
+	union {
+		struct azure_iot_hub_data msg;
+		struct azure_iot_hub_method method;
+		struct azure_iot_hub_result result;
+		int err;
+		bool persistent_session;
+		uint16_t message_id;
+	} data;
+	/* Associated topic, not relevant for all events. */
+	struct azure_iot_hub_topic_data topic;
+};
+
+/** @brief Azure IoT Hub library event handler.
+ *
+ *  @param evt Pointer to event structure.
+ */
+typedef void (*azure_iot_hub_evt_handler_t)(struct azure_iot_hub_evt *evt);
+
+/** @brief Structure for Azure IoT Hub broker connection parameters. */
+struct azure_iot_hub_config {
+	/** Socket for Azure IoT Hub broker connection. Polling on the socket
+	 *  happens internally in the library.
+	 */
+	int socket;
+	/** Device id for Azure IoT Hub broker connection.
+	 *  If NULL, the device ID provided by Kconfig is used.
+	 */
+	char *device_id;
+	/** Length of device string. */
+	size_t device_id_len;
+};
+
+/** @brief Initialize the module.
+ *
+ *  @warning This API must be called exactly once, and it must return
+ *           successfully for subsequent calls to this library to succeed.
+ *
+ *  @param[in] config Pointer to struct containing connection parameters. Can be
+ *		      NULL if CONFIG_AZURE_IOT_HUB_DEVICE_ID is set.
+ *  @param[in] event_handler Pointer to event handler function.
+ *
+ *  @return 0 If successful. Otherwise, a (negative) error code is returned.
```

Comment:
```suggestion
 *  @retval 0 If successful. Otherwise, a (negative) error code is returned.
```

Reply:
My understanding is that `retval` is to be used if an exhaustive list of all possible return values is provided, and `return` in case of a textual description like this.
@ru-fu @umapraseeda Any input on this?